Add initial integration test setup and first few tests

jrfnl opened the pull request:
## Introduction

This PR creates an initial test framework to allow for creating and running integration/end-to-end tests for this plugin.

The framework as I have set it up:
* Allows for running tests both as a "project local" as well as a Composer `global` install _without screwing up the Composer global environment of the user running the tests_.
    Tests are run in the system temp directory and the tests clean up after themselves.
* Will ensure that the tests will run with the plugin _as it is in the branch being tested_.
* As the tests start new PHP processes, the framework ensures that the PHP version used to _initiate the test run_ will be used in those processes. This prevents a test run on a dev machine using the system default PHP version, which would skew the test results.
* Similarly, the framework can be told to use a specific Composer phar file to run the tests with, rather than the system default version.
* The framework also allows for setting up "fake" PHPCS standards for use in the tests, which will make writing the tests simpler as we don't need to keep track of changes in external standards and don't need to take those into account in the tests.
* Tests can be run against multiple PHPCS versions using data providers and a helper class is included to get a specific or random) selection of versions to use compatible with the PHP version against which the tests are being run.
* And to aid in debugging failing tests, it includes a `TestListener` which will spit out all the CLI output seen for a particular test when the test fails.

The tests included in this PR are largely the tests which were already previously being run via the GH Actions workflow, though they are now encapsuled within a PHPUnit based framework and will be run via GH Actions against all supported PHP version, multiple versions of Composer and on multiple OSes.

A lot more tests are still needed, but those should be relatively straight-forward to add once this initial framework is in place and I have already started preparing some follow-up PRs to be pulled once this PR has been merged.

⚠️  This PR depends on PRs #147 and #152 both being accepted and merged.

👉🏻 As this is a huge PR, I have split it up into atomic commits telling the "story" and I would strongly recommend reviewing this PR by looking at each commit separately.

## Proposed Changes

### Tests: create initial setup

* Require the PHPUnit Polyfills as a basis for the tests.
    The PHPUnit Polyfills will ensure a compatible PHPUnit version will be available and will allow for writing tests for the latest version of PHPUnit (9.x), while still being able to run the tests on PHPUnit 4-8.
* Include a basic test bootstrap file which will allow for running the tests both via a Composer installed version of PHPUnit as well as via a PHPUnit PHAR file.
    Note: the bootstrap file will try to determine where the `composer.phar` file is located. For running the tests locally and/or overruling the default version of Composer, a `COMPOSER_PHAR` environment variable can be set in a local `phpunit.xml` overload file.
* Include information about the test setup in the `CONTRIBUTING` documentation.
* Includes adding new jobs to the "Quick test" and "Integration test" GH Actions workflows.
    Once the existing tests in the _old_ workflows have been converted to the new test setup, the old workflows can be removed.
* Minor tweaks to various repository config files.

Note: this test setup presumes that the proposal to drop support for PHP 5.3 - see issue #145 - will be accepted as the PHPUnit Polyfills require PHP >= 5.4.

Refs:
* https://github.com/Yoast/PHPUnit-Polyfills/

### TestCase: add helper methods for setup/teardown

As most tests will need to run a `composer install`, tests should be run in separate, temporary directories.

With that in mind, I'm adding two helper methods to the base `TestCase` class:
* `createTestEnvironment()` - this method will create a `PHPCSPluginTest/TestClassName_uniqueID/local` and a `PHPCSPluginTest/TestClassName_uniqueID/global` directory in the system temp directory and set the Composer global `HOME` directory to the created `global` directory.
* `removeTestEnvironment()` - this method will remove the created directories and reset the Composer global `HOME` directory to its original location.

Notes:
* The paths to the created directories will be added to the `$tempDir`, `$tempGlobalPath` and `$tempLocalPath` properties of the class and can be used in the tests.
* The methods (and associated properties) have been made `static` on purpose to allow for these methods to be used both in `setUp()`/`tearDown()` fixtures as well as in `setUpBeforeClass()`/`tearDownAfterClass()` fixtures.
    The latter would be useful for a test class needing only one environment with various test methods depending on each other and building on each other.
    This is also the reason that these methods are set up as helpers instead of fixtures, as this way, the decision what fixtures are most appropriate can be made for each individual concrete test class.
* A hard cleanup of the `PHPCSPluginTest` directory in the system temp directory has been added to the test bootstrap to make sure that any stray "old" test directories will be removed before a new test run.

Includes a `static` `TestCase::onWindows()` helper method to determine whether or not the tests are being run on Windows.

### TestCase: add helper method for on-the-fly composer.json creation 

This adds a test helper method, which, given an array with a Composer configuration and a target location, will on-the-fly create a `composer.json` file for use in a test.

The method will automatically add two additional settings to the `composer.json` file:
* It will add a `install-codestandards` script to run the plugin on demand.
* It will add the `allow-plugins` permission, which is needed for Composer 2.2.

By adding these keys automatically, this kind of "noise" is not needed in the actual test classes.

The method is `static` so as to allow for it to be used in `setUpBeforeClass()`/`tearDownAfterClass()` fixtures.

### Tests: create zip package artifact for the current plugin code

In normal circumstances, a `composer install` will download a version of the plugin from Packagist.
This will not work for the tests however, as the development version of the plugin being tested will generally not be available on Packagist.

With this in mind, we need a work-around to ensure that the tests are run with the **current** version of the plugin as per the last commit in the current branch.

I've implemented this by using the `repositories - artifact` feature of Composer.

This works as follow:
* In the test bootstrap, ahead of the test run, the plugin is zipped up as a package and placed in the `tests/artifact` directory.
* Whenever a `composer install/update` command is run within the test suite, the `repositories-artifact` key is injected into the `composer.json` when the `composer.json` file is written to the temp directory used for the test.
* With this in place, Composer will install the plugin from the zipped file in the `tests/artifact` directory instead of downloading from Packagist.

Notes:
* The version number to be used for the zip artifact of the plugin is defined in the test bootstrap file.
    This number may need updating once in a while to ensure it is higher than the latest released version of the plugin.
* In the `CreateComposerZipArtifacts` class, three properties are used to determine which files to exclude from the package - `$excludedFiles`, `$excludedExtensions` and `$excludedDirs`.
    Depending on new files/directories being introduced in the package, these lists may need to be updated once in a while.

Ref: https://getcomposer.org/doc/05-repositories.md#artifact

### TestCase: add helper methods to run CLI commands and custom assertions

This commit adds six additional methods to the base `TestCase` class:
* `executeCliCommand()`: a helper method to execute a CLI command and return the exit code, the contents of `stdout` and the content of `stderr` for use in assertions.
* `stabilizeCommand()`: a helper method which is automatically called from the `executeCliCommand()` method. This method will:
    - Ensure that the PHP version which was used to initiate the test run will also be used in the CLI command.
        Without this, the system default PHP version would be used for CLI commands, which may be a different version from the PHP version used to run the tests and would skew the test results.
    - Ensure that the Composer version as set up using the `env` variable is used in the tests.
        Again, this will prevent the system default version of Composer being used, which may skew the results when the tests are intended to be run against a specific version of Composer.
    - Adds the `--no-interaction` flag to all Composer CLI commands to prevent tests hanging when interaction would be expected.
    - Adds quotes around a command when the tests are run on Windows in combination with PHP < 8 to ensure cross-platform compatibility of the CLI commands.
* `getPhpcsCommand()`: a helper method to get the path to the vendor installed PHPCS entry point file.
* `maybeStripColors()`: a helper method to compare CLI output. **This method is still WIP and will probably still need some tweaking.**
    This method will take an output expectation and an actual output and will attempt to strip CLI colour codes from the actual output when the output expectation does not contain colour codes. This way, expected colour codes can still be tested by including them in the output expectation.
    Note: until the bugs in this method have been smoothed out, it is recommended to use `--no-ansi` CLI arguments in CLI commands not testing for colour codes.
* `assertExecute()`: a custom assertion which will run a given CLI command, optionally in a specific working directory and will subsequently assert that the exit code is the same as expected and the `stdout` and `stderr` outputs contain the expected output.
    For more complex assertions against the output of CLI commands, the `executeCliCommand()` can be used directly in tests.
    The `StringContainsString` assertions used for `stdout` and `stderr` will attempt to strip color codes from the output before doing a compare using the `maybeStripColors()` method.
* `assertComposerValidates()`: a custom assertion to verify that a given `composer.json` file validates for use in the tests.

A number of these methods have been made `static` to allow for them to be used in conjunction with `setUpBeforeClass()` fixtures.

### Tests: add TestListener to log and display CLI output for failing tests

While not necessarily the most pretty solution (output is displayed in the middle of the progress report), this allows for access to the `composer.json` content, the CLI commands and their output for any test which failed, which will be helpful when debugging test failures, especially in CI.

### Tests: add initial baseline test

These tests verify:
- That the plugin can be installed and functions correctly with the full range of supported PHPCS versions.
    While the test is not run against the full range of supported PHPCS version each time, due to the random PHPCS version selection, all supported PHPCS versions will be tested over time in CI.
- That the plugin runs when installed.
- That the PHPCS native standards are the only recognized standards when no external standards are available.
- That no `CodeSniffer.conf``file gets created when no external standards are found.

These tests will be run in both a Composer global as well as a Composer project local environment.

The tests use a dataprovider which will randomly select two PHPCS versions compatible with the PHP version on which the test is being run + PHPCS `dev-master` + PHPCS `4.x` (dev version for the next major).

As which PHPCS versions are compatible with which PHP versions needs quite some logic, a separate helper class `PHPCSVersions` is introduced, which encapsulates that knowledge.

The `PHPCSVersions` class contains a number of (`static`) methods as helpers. These can be used in any test class:
* `get()`: retrieves an array with a specific number of PHPCS versions valid for the current PHP version.
* `getHighLow()`: retrieves an array of the highest and lowest PHPCS versions valid for the current PHP version.
* `getHighLowEachMajor()`: retrieves an array of the highest and lowest supported PHPCS versions for each PHPCS major (and valid for the current PHP version).
* `getRandom()`: gets a random PHPCS version which is valid for the current PHP version.
* `toDataprovider()`: converts a version array to an array suitable for use as a PHPUnit dataprovider.
* `getSupportedVersions()`: retrieves an array with all PHPCS versions valid for the current PHP version.
* `isNextMajorSupported()`: determines whether the current PHP version is supported on the "next major" branch of PHPCS.
* `getStandards()`: retrieves an array of the names of the PHPCS native standards for a particular PHPCS version.

Generally speaking, the `dev-master` and `4.x` branch will not be included in the retrieved ranges, but can be specifically requested by passing optional parameters.

This `PHPCSVersions` class will need semi-regular updates when new versions of PHPCS are released and new PHP versions are released.

### Tests: move test with external standard from CI to test class

This moves the test which was being run via GH Actions to a test class in the integration test suite.

Differences between the previous and new setup:
* The previous test via GH Actions would only fail on the exit code not being `0` for any of the steps.
    The new PHPUnit based tests executes more detailed assertions on the results of the commands being run in the tests.
* The previous test would execute the test run of PHPCS against the `src/Plugin.php` file in this package.
    The new PHPUnit based tests will execute against a simple PHP file created on the fly.
* The previous test via GH Actions would only run for a Composer `local` situation.
    The new PHPUnit based tests cover both a Composer `global` as well as a `local` setup.
* The matrix against which the tests are run is different.
    In the old situation, the test would only be run on a fixed set of PHP/PHPCS combinations with Composer 2.x on Ubuntu.
    In the new situation, the tests will run against both Composer 1.x, as well as 2.x on both Ubuntu and Windows.
    Same as before, it will be run against all supported PHP versions (in the extended `integrationtest` workflow), but the PHPCS versions to test against are more varied and - in part - randomly selected out of the PHPCS version compatible with the current PHP version, which results in more comprehensive testing of these scenarios.

Includes:
* Adding a new `createFile()` helper method to the base `TestCase` to generate a (PHP) file to do a test run against.
* Removing the "old" integration test from the GH Actions script.

### Tests: set up a fake PHPCS standard as a fixture and create an artifact of it

The tests in the `RegisterExternalStandardsTest` class contain quite some work-arounds to handle changes made over time to the external PHPCompatibility standard.

The `PHPCompatibility` standard was chosen for use in the tests as it is one of the rare standards which still supports both PHPCS 2.x as well as 3.x.

All the same, to improve the stability of the tests and to simplify the actual test code, using "fake" PHPCS standards set up as fixtures as part of this test suite will make life easier.

This commit creates the initial setup for that, by:
* Adding code to the `CreateComposerZipArtifacts` class to handle creating Composer package zip artifacts for each subdirectory of the `tests/fixtures` directory.
* Calling that code from the test bootstrap file to ensure all test artifacts are zipped up and ready for use ahead of running the tests.
* Adding an initial "fake" `DummySubDir` standard as a test fixture.
* Adding a `README` file to the `tests/fixtures` directory with the rationale and a list of available fixtures.

### RegisterExternalStandardsTest: switch out PHPCompatibility for the test fixture

This replaces the use of the PHPCompatibility standard in the `RegisterExternalStandardsTest` tests with the use of the `dummy-subdir` fake standard fixture.

### RegisterExternalStandardsTest: simplify check whether PHPCS can run with the standard

To check whether PHPCS can run with the standard, previously, a scan on a simple file was done, with this file being created on the fly for each test.

This commit replaces that check with running the "explain" command.

Advantages:
* The "explain" command does not need a file to scan and can still confirm that the standard can be read correctly by PHPCS.
* The output for the "explain" command has fewer differences across PHPCS versions, so is simpler to verify.


### :new: BaseLineTest: selectively skip the tests

... in a very specific combination of circumstances - Windows + Composer 1.x + PHP 5.5 + PHPCS `dev-master` + no external standards -, as the Composer logs in that case, don't always show the "Running PHPCodeSniffer Composer Installer" message.

As it is such a specific combination, which will probably be rare to encounter in real life, I'm proposing to skip the tests for that combination instead of spending lots of time trying to debug this (for now).

### 🆕 TestCase: add `willPluginOutputShow()` method

In very select circumstances (PHP 5.5 with Composer 1.x on Windows) and sometimes even only with select PHPCS versions, the plugin output will not be shown in the transcript from Composer, even though the plugin _does_ actually run.

Instead of skipping those tests completely, I'm suggestion to selectively skip the output expectation for the `composer install/update` run instead.
That way we can still verify that the plugin _has_ actually done its job by checking that paths have been registered with PHPCS.

I would recommend to only implement the use of this method for those select tests where we have actually seen failures due to this Composer bug and to just leave the other tests alone.

## Related Issues

* https://github.com/Dealerdirect/phpcodesniffer-composer-installer/issues/9
* https://github.com/Dealerdirect/phpcodesniffer-composer-installer/issues/92

Closes #9

jrfnl commented:
Note: the test failure on PHP 5.5 with Composer v1 on Windows is a really weird one. It doesn't fail on every run, just most of the time and it only seems to fail for PHPCS `dev-master`, which makes this even weirder. I've been trying to figure out the cause, but haven't been able to get to the bottom of it so far.

As it is a very specific combination of circumstances and one which will probably be rare nowadays, we may need to decide to just skip the test for that specific combination.

Potherca commented:
Review planned on: <kbd>2022-02-04</kbd>.

jrfnl commented:
> Note: the test failure on PHP 5.5 with Composer v1 on Windows is a really weird one. It doesn't fail on every run, just most of the time and it only seems to fail for PHPCS `dev-master`, which makes this even weirder. I've been trying to figure out the cause, but haven't been able to get to the bottom of it so far.
> 
> As it is a very specific combination of circumstances and one which will probably be rare IRL nowadays, we may need to decide to just skip the test for that specific combination.

I've added a commit to skip the test which fails intermittently for now.

jrfnl commented:
Note: to make this PR mergable, the required checks for the branch (branch protection settings) will need updating as well.

jrfnl commented:
I've made a minor update to the `PHPCSVersions` class to ensure it is in-line with the update I made to PR #152.

jrfnl commented:
Rebased without changes to get it past the merge conflicts.

jrfnl commented:
Oh joy.. the new assignment operator alignment sniff was throwing some issues. Fixed up now in the respective commits. No changes other than those whitespace fixes.

jrfnl commented:
FYI: push was to make a tiny formatting tweak to the `README` for the fixtures, no textual or functional changes.

jrfnl commented:
Added one more commit to the PR as I found that some of the other tests I'm working on will also need to selectively skip a certain output expectation. Abstracting the condition to a method will make it more easily re-usable in those other tests.

Potherca commented on tests/CreateComposerZipArtifacts.php:
```diff
@@ -0,0 +1,259 @@
+            $artifactDir .= '/';
+        }
+
+        $this->artifactDir = $artifactDir;
```
To my mind, this would be more readable:
```php
        // Make sure the directory has a trailing slash.
        $this->artifactDir = rtrim($artifactDir, '/') . '/';
```

Potherca commented on tests/CreateComposerZipArtifacts.php:
```diff
@@ -0,0 +1,259 @@
+                continue;
+            }
+
+            @unlink($fileinfo->getPathname());
```
Wouldn't a positive match be more logical here?
```php
        foreach ($di as $fileinfo) {
            if ($fileinfo->isFile() && $fileinfo->getExtension() === 'zip') {
                @unlink($fileinfo->getPathname());
            }
        }

```

Potherca commented on tests/IntegrationTest/BaseLineTest.php:
```diff
@@ -0,0 +1,167 @@
+     */
+    public function testBaseLineGlobal($phpcsVersion, $expectedStnds)
+    {
+        $config                                             = $this->composerConfig;
```
This is one of those places where aligned values read weird to me, because of the large gap.

jrfnl replied:
I agree. I was actually surprised the sniff demanded it as I though the limit was smaller.

Let me see if I can configure the limit in the ruleset to sort this out better.

jrfnl replied:
Done.

Potherca commented on tests/IntegrationTest/BaseLineTest.php:
```diff
@@ -0,0 +1,167 @@
+        $config                                             = $this->composerConfig;
+        $config['require-dev']['squizlabs/php_codesniffer'] = $phpcsVersion;
+
+        $this->writeComposerJsonFile($config, static::$tempGlobalPath);
```
As the BaseLineTest class is `final` (and we don't expect the class to be extended anyway), `self` can be used instead of `static`.

jrfnl replied:
True, it's more of a consistency thing to always use `static::` unless I explicitly want `self` or `parent`.

I have no objections to changing it though (though will need to change it everywhere as I have used it everywhere).

Potherca replied:
Its fine to leave as-is. Its just something that stood out when reviewing.

Potherca commented on tests/IntegrationTest/RegisterExternalStandardsTest.php:
```diff
@@ -0,0 +1,191 @@
+     *
+     * @return array
+     */
+    public function dataRegisterOneStandard()
```
Keeping with the "Use a verb for method names", I'd suggest starting data provider methods with `provide` (similar to test methods starting with `test`).

jrfnl replied:
Ha, two different schools of conventions clashing here.

I'm used to test data providers always being prefixed with `data`.

I'm generally in the habit of naming data providers which are linked to one test the same as the test with `test` replaced by `data`.
When a data provider is used by multiple tests, a generic, descriptive name, but again prefixed with `data`, is used.

Potherca replied:
I tend to name the providers `provide...` with a description of what they are providing, to make the purpose of the data more clear to those unfamiliar with the inner workings of the code. I falls in line with `test...`, `mock...` and `assert...`.

But yeah, if the worst thing I can find in the code is the naming of methods, the code itself is pretty good.

jrfnl replied:
Thanks ;-)

I generally make sure that the test methods which use data providers have a docblocks with full `@param` annotations, that the parameter names are descriptive and then have the data provider closely coupled by placing it directly following the test method and using named data providers, with each case having a useful name to know what's being tested + using the parameter names in the sub arrays as indexes as well.

Altogether it might be a bit over the top, but I would be very surprised if there could ever be confusion about the purpose of the data with a setup like that... 😇

Potherca commented on tests/fixtures/README.md:
```diff
@@ -0,0 +1,42 @@
+| Characteristics          | Notes                                                                                            |
+|--------------------------|--------------------------------------------------------------------------------------------------|
+| **Standard(s):**         | `DummySubDir`                                                                                    |
+| **Includes sniff(s):**   | :heavy_checkmark: One sniff - `DummySubDir.Demo.Demo` - which is PHPCS cross-version compatible. |
```
To make sure the text renders across different tools/stacks, I'd suggest using the actual emoji rather than the word mark.

jrfnl replied:
I've never actually heard of problems with the rendering... can you give me an example ?

Potherca replied:
Basically any places that does not do `:..:` parsing, like text-editors, command-line, etc.

![image](https://user-images.githubusercontent.com/195757/152693479-da93607d-a066-4972-90b9-2360898aecd9.png)

jrfnl replied:
Ah, but in my mind, that's a good thing as a) not all editors support UTF-8, b) the file may not have been saved as UTF-8 to begin with, which could lead to a corrupted character and c) rendering of emojis on different OS-es or even different websites can be different, so having the original intention in the text by using the `:...:` annotation makes it clear to whomever is editing what the intended emoji is.

Potherca reviewed (approved):
I've got some questions and some pointers but nothing that would stop this from being merged.

Once I'd reviewed the bootstrap and testcase class, the actual test were easy to read/understand. :+1:

jrfnl commented on .github/workflows/integrationtest.yml:
```diff
@@ -151,41 +53,27 @@ jobs:
-        if: ${{ matrix.phpcompat == 'composer' }}
-        run: ./vendor/bin/phpcs -ps ./src/ --standard=PHPCompatibility --sniffs=PHPCompatibility.FunctionUse.RemovedFunctions --runtime-set testVersion ${{ matrix.php }}
+      - name: Run integration tests
+        run: vendor/bin/phpunit --no-coverage
```
I understand where you are coming from, but I've generally found it a good thing to keep code in line with the current reality, without making accommodations for future realities which may never come to be.

To put more plainly: if unit tests would ever be added, that would be the time IMO to adjust the workflows to be more selective, not now.

I should also point out that adding unit tests to this code base would be neigh impossible, what with nearly all methods being `private`.

jrfnl commented on CONTRIBUTING.md:
```diff
@@ -57,13 +60,48 @@ can be downloaded suitable for your operating system from their [releases page][
+    **Note**: this setting also allows for locally testing with different versions of Composer.
+    You could, for instance, have multiple Composer Phar files locally, `composer1.phar`, `composer2.1.phar`, `composer2.2.phar`.
+    By changing the path in the value of this `env` setting, you can switch which version will be used in the tests.
+
```
Which exception cases are you talking about ? The Windows/PHP 5.5/Composer 1.x issue ?

There are a number of different things the tests (and sometimes the plugin too) take into account for cross-version compatibility.
In the tests I've made those situation as clear as possible and added ample inline/docblock documentation about these situations.

Potherca replied:
Agreed, the comments in the test are sufficient.

jrfnl commented on README.md:
```diff
@@ -49,7 +49,7 @@ That's it.
 This plugin is compatible with:
 
-- PHP **5.x**, **7.x**, and **8.x** (Support for PHP v8 is available since [`v0.7.0`][v0.7])
+- PHP **5.4+**, **7.x**, and **8.x** (Support for PHP v8 is available since [`v0.7.0`][v0.7])
```
That's something for the changelog IMO and should be handled in PR #147 (where this change comes from).

jrfnl commented on composer.json:
```diff
@@ -27,14 +27,16 @@
     "phpcompatibility/php-compatibility": "^9.0",
-    "php-parallel-lint/php-parallel-lint": "^1.3.1"
+    "php-parallel-lint/php-parallel-lint": "^1.3.1",
+    "yoast/phpunit-polyfills": "^1.0.1"
```
Good point, `^1.0` would suffice.

The difference with `1.0.1` is that `1.0.1` has a publicly accessible version number available, which _could_ come in handy, though I've not needed it yet.

Ref: https://github.com/Yoast/PHPUnit-Polyfills/releases

jrfnl commented on composer.json:
```diff
@@ -52,6 +59,12 @@
       "@php ./vendor/php-parallel-lint/php-parallel-lint/parallel-lint . -e php --exclude vendor --exclude .git"
+    ],
+    "test": [
+      "@php ./vendor/phpunit/phpunit/phpunit --no-coverage"
```
Using `phpunit` here is generally a bad idea as if there is a `phpunit` executable in the system path before the path to Composer, that will often take precedence.

Using `vendor/bin/phpunit` is also not a good idea as then the PHP version used by Composer won't be respected.

Using `@php vendor/bin/phpunit` would work for *nix systems, but breaks on Windows.

And that's how you end up with `@php ./vendor/phpunit/phpunit/phpunit` which:
* Respects the PHP version used by Composer
* Will always use the Composer installed version of PHPUnit rather than whatever system default is set up
* Will work cross-platform.

Potherca replied:
Cool. Thanks for the details!

jrfnl commented on phpunit.xml.dist:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/7.5/phpunit.xsd"
```
Well, the current schema used should be compatible with PHPUnit 4.x - 9.2. The schema changed in PHPUnit 9.3, but using the new schema is (obviously) not an option if we want the tests to keep running on older PHPUnit versions.

Once PHPUnit 10.x comes out, the CI may need the "on-the-fly" update of the XML file using `--migrate-configuration`.

So with all this in mind, I just went with a version which is sort of in the middle of all supported PHPUnit versions.

Re: using the online scheme vs the one in vendor - the one in `vendor` will change depending on the PHP version used to run `composer install` as the most appropriate PHPUnit version will be installed, so that doesn't feel like a reliable solution.

Potherca replied:
Agreed.

jrfnl commented on tests/PHPCSVersions.php:
```diff
@@ -0,0 +1,425 @@
+     */
+    public static function getSupportedVersions()
+    {
+        $versions = self::$allPhpcsVersions;
```
Good catch! I change the logic in that function a couple of times and forgot to remove that line 👍🏻

jrfnl commented on tests/PHPCSVersions.php:
```diff
@@ -0,0 +1,425 @@
+        if (
+            is_string($version) === false
+            || (isset(self::$allPhpcsVersions[$version]) === false
+            && $version !== PHPCSVersions::MASTER
```
👍🏻 

Copy/pasta from when I originally created the function in one of the test files and then later decided to move it to this class.

jrfnl commented on tests/PHPCSVersions.php:
```diff
@@ -0,0 +1,425 @@
+            $standards[] = 'PSR12';
+        }
+
+        sort($standards, \SORT_NATURAL);
```
`natsort()` preserves the keys in the array, `sort(..., SORT_NATURAL)` will reindex the array for numerically indexed arrays.

As this method is generally used as input for expectations to be fed to `Assert::assertSame()`, the keys will be part of the comparison, so always re-indexing them for arrays where the keys are not important allows for those assertions to pass.

Refs:
* https://www.php.net/manual/en/function.natsort.php
* https://www.php.net/manual/en/function.sort.php
* https://www.php.net/manual/en/array.sorting.php

Potherca replied:
> natsort() preserves the keys in the array, sort(..., SORT_NATURAL) will reindex the array for numerically indexed arrays.

Ah. I did not know that. Cool!

jrfnl commented on tests/IntegrationTest/BaseLineTest.php:
```diff
@@ -0,0 +1,167 @@
+            'Failed to install dependencies.'
+        );
+
+        $result = $this->executeCliCommand('"vendor/bin/phpcs" -i', static::$tempGlobalPath);
```
If the method is called from within an instantiated class which extends `TestCase`, it doesn't matter, all three (four if we also count `parent::`) ways of calling the method are perfectly valid.

It only matters when the method is used outside of an instantiated class. In that case, only `self::`, `parent::` or `static::` are available as `static` methods cannot be called non-statically outside of an instantiated class.

The `executeCliCommand()` method was made static to allow it to be used in `setUpBeforeClass()` methods too and when used in those, it can only be called statically.

On that note - the same can be said about the PHPUnit native/polyfilled assertion calls as those are all defined as static methods too, but commonly called using `$this->assert...()` (including in this test suite).

jrfnl commented on tests/TestCase.php:
```diff
@@ -0,0 +1,582 @@
+     */
+    protected static function onWindows()
+    {
+        return strpos(strtoupper(\PHP_OS), 'WIN') === 0;
```
Oh, good one. Copy/pasta from elsewhere without critically looking at it (though there is good reason for doing the comparison case-insensitively).

jrfnl commented on tests/TestCase.php:
```diff
@@ -0,0 +1,582 @@
+         * Disable TLS when on Windows with Composer 1.x and PHP 5.4.
+         * @link https://github.com/composer/composer/issues/10495
+         */
+        if (static::onWindows() === true && \CLI_PHP_MINOR === '5.4' && substr(\COMPOSER_VERSION, 0, 1) === '1') {
```
*grin* I originally wrote it as the second option, but then remembered that that is PHP 5.6+ syntax and not supported on PHP 5.4.

As for the first option, no particular reason other than `substr(\COMPOSER_VERSION, 0, 1) === '1'` being slightly more obvious what it's doing than `strpos(\COMPOSER_VERSION, '1') === 0`.

jrfnl commented on tests/TestCase.php:
```diff
@@ -0,0 +1,582 @@
+         * Note: it is unclear what changes in PHP 8.0, but the quotes will now suddenly break things.
+         * Ref: https://www.php.net/manual/en/function.proc-open.php#example-3331
+         */
+        if (static::onWindows() === true && substr(\CLI_PHP_MINOR, 0, 1) < 8) {
```
... but unfortunately is PHP >= 5.6 syntax, so not usable for this test suite.

Potherca replied:
I know `$string{0}` was deprecated somewhere along the way (PHP 7.4, I think?) but I thought array-style access for string index always just worked? :raised_eyebrow: 

Hmmm... According to https://3v4l.org/2f0CM everything seems fine.

jrfnl replied:
> I know `$string{0}` was deprecated somewhere along the way (PHP 7.4, I think?) but I thought array-style access for string index always just worked? 🤨

On variables, not on constants.

> Hmmm... According to https://3v4l.org/2f0CM everything seems fine.

Eh.. no it isn't. https://3v4l.org/rJWEX

jrfnl commented on tests/TestCase.php:
```diff
@@ -0,0 +1,582 @@
+        }
+
+        if (
+            strpos($expected, "\033") === false && strpos($actual, "\033") !== false
```
Thanks for catching that!

Note: that method still needs more work, but it's a start anyway.

jrfnl commented:
Thanks for the extensive review @Potherca ! I've left replies everywhere relevant and adjusted those things for which there is no discussion.

Potherca commented:
As all my points have been addressed and all questions have been answered, this is good to merge if you are content with it. :+1:

jrfnl commented:
> As all my points have been addressed and all questions have been answered, this is good to merge if you are content with it. 👍

I am. ✔️ 

I do have some follow-up PRs with tweaks, but didn't want to make the review of this PR harder than it would already be, so I will pull those changes + additional tests in separate PRs once this PR has been merged.

You can expect some ~15 follow-up PRs based on what I have lined up locally so far.

> Does #147 have to merged first?

It does. As 0.7.2 has been released now, I've moved that one from draft to "ready for review". Once #147 has been merged, I'll mark this as ready (and if needs be rebase it without changes to get past potential imaginary merge conflicts).

The branch protection settings will need updating for both PRs, so may be easiest to just do that once after this one has been merged.

@mjrider Did you still want to have a look at either of these PRs before we merge them ?

jrfnl commented:
> This PR depends on PRs https://github.com/PHPCSStandards/composer-installer/pull/147 and https://github.com/PHPCSStandards/composer-installer/pull/152 both being accepted and merged.

Rebased now the prerequisites have both been merged & marking this ready for review/merge.

jrfnl commented:
> Note: to make this PR mergable, the required checks for the branch (branch protection settings) will need updating as well.

I've updated the branch protection settings to match the new builds and removed the old builds. PR should be mergable now.

Potherca commented:
@jrfnl Random musings: Do we want to skip v0.8 and just release [all open changes](https://github.com/PHPCSStandards/composer-installer/compare/v0.7.2...master) together with this MR as v1 ?

jrfnl commented:
Thanks for getting this merged. I'll start pulling the follow up PRs.

jrfnl commented:
Related to #92
